### ISSUE TITLE
cpp_new.com terminates halfway

### ISSUE THREAD

**Issue by ostnils** [nikitinprior/dcpp#1]
This testfile
#include "stdio.h"
#include "t.h"
int
main(int argc, char **argv) {

  fprintf(stderr, "XYZ is \"%s\"\n", XYZ);
}
and t.h is
#define XYZ "abc"
i = 3
the only output when I run it with 
zxcc cpp_new.com t.c
the only output are:
 1 "p:t.c"

include

My environment: Mint Linux, ZXCC from agn453/ZXCC and Hi-Tech C from agn453/Hi-Tech-Z80-C.

Don't hesitate to contact me if you need more information
/anders

**Comment by nikitinprior** [comment]
Only four programs zxc (equivalent to C.COM), zxas (equivalent to ZAS.COM), zxlink (equivalent to LINK.COM) and zxlibr (equivalent to LIBR.COM) convert their arguments to a form suitable for zxcc and then call zxcc.

For other CP/M programs, you have to enter the arguments for them in the zxcc format specified in the zxcc documentation.

Best regards,
Andrey

**Comment by ostnils** [comment]
Yes I know that the xc* programs are the one to use. But the problem still
exists. I hacked the c.com program to not delete the output from cpp.com
(the $ctmp1.$$$ file). When I replaced cpp.com from agn453's distribution
with cpp_new (built with your makefile (have to change the name of the
linker to 'linq' in the makefile) and agn453 dist) the following was the
obviously truncated output from cpp (see below).

If you are curious about my interest in your version of cpp, I need your
-noinclude flag, I just started to port unix V7 to a Zilog z280 based board
(cpu280 if you hear about it). The problem is that the original cpp
includes Hi-Techs include files. Hi-tech's stat.h do not work well with the
V7 sources:(

/anders

The output from vprintf.c produced by cpp_new.com (I let all the whitespace
be left. The file ends after the 'r', no LF or CR

# 1 "P:VPRINTF.C"





# 1 "C:stdio.h"









# 13 "C:stdio.h"





extern struct _iobuf
{
    char *_ptr;
    int _cnt;
    char *_base;
    unsigned short _flag;
    char _file;
} _iob[ 8];










































# 71 "C:stdio.h"














extern int fclose( struct _iobuf *);
extern int fflush( struct _iobuf *);
extern int fgetc( struct _iobuf *);
extern int ungetc(int, struct _iobuf *);
extern int fputc(int, struct _iobuf *);
extern int getw( struct _iobuf *);
extern int putw(int, struct _iobuf *);
extern char *gets(char *);
extern int puts(char *);
extern int fputs(char *, struct _iobuf *);
extern int fread(void *, unsigned, unsigned, struct _iobuf *);
extern int fwrite(void *, unsigned, unsigned, struct _iobuf *);
extern int fseek( struct _iobuf *, long, int);
extern int rewind( struct _iobuf *);
extern int setbuf( struct _iobuf *, char *);
extern int printf(char *, ...);
extern int fprintf( struct _iobuf *, char *, ...);
extern int sprintf(char *, char *, ...);
extern int scanf(char *, ...);
extern int fscanf( struct _iobuf *, char *, ...);
extern int sscanf(char *, char *, ...);
extern int remove(char *);
extern struct _iobuf *fopen(char *, char *);
extern struct _iobuf *freopen(char *, char *, struct _iobuf *);
extern struct _iobuf *fdopen(int, char *);
extern long ftell( struct _iobuf *);
extern char *fgets(char *, int, struct _iobuf *);
extern char *_bufallo(void);


# 6 "P:VPRINTF.C"

# 1 "C:stdarg.h"





typedef void * va_list[1];








# 7 "P:VPRINTF.C"

extern int _doprnt();

int
vfprintf( struct _iobuf * file, char * fmt, va_list args) {
r

Den mån 21 nov. 2022 kl 21:28 skrev nikitinprior ***@***.***>:

**Comment by nikitinprior** [comment]
I am interested in your wish and I have a few questions for you.

However, it is inconvenient for me to answer you via issues github.com .

It is better to exchange information by e-mail.
My email address is listed in the program description in the repositories.

**Comment by nikitinprior** [comment]
The Unix Edition 7 operating system is written in C and for this reason it can be transferred to a microcomputer with a different processor.
To do this, you need a compiler that can translate the source code in C into the commands of the desired microprocessor. However, not all programs can be compiled with the new processor. Part of the operating system code depends on the hardware of the new microcomputer and the existing machine-dependent Unix Edition 7 code is not suitable for this.
Since the existence of the Unix operating system, several attempts have been made to transfer it to the z80 processor. Micronix migrated Unix V6. They used Whitesmiths C as a compiler.
Douglas Braun wrote a Unix-like operating system UZI using Code Works Q/C.
Harold F. Bower adapted UZI into UZI180 using HiTech C v3.09. Stefan Nitschke developed it into UZI280, also using HiTech C v3.09.
During the development of these operating systems, only a small part of microcomputers could execute Unix code. At that time , MSX microcomputers and
Arcady Schekochikhin and Adriano C. R. da Cunha developed UZIX. First v1 and then V2. At that point in time, they had advanced the furthest. However, their operating system only supported MSX hardware. In addition, the source code of UZIX v2 has not been published.
Sergey Akimov adapted UZIX for the ORION microcomputer.
Alan Cox started the development of the FUZIX operating system.
He now has the most advanced operating system that can run on different processors and different microcomputers. FUZIX is based on the UZI180 and complements it with a number of missing features. However, as a compiler, he used SDCC, which is a cross-compiler and cannot work on processors for which he generates code. This is a big drawback of FUZIX.
Writing a multi-task and multi-user Unix-like operating system is quite a difficult task. Many started it and as a result could not bring it to the end. The reasons were different, but the result was sad for everyone. It is almost impossible for one person to do such work qualitatively. Since at this point in time there are few computers left for old processors and even fewer people who use them, testing the operating system becomes quite difficult. The fewer users there are, the less likely it is to detect errors and fix them. And a system with errors is not very attractive to new users. It's a vicious circle.
Your desire to port Unix Edition 7 to z280 causes respect on the one hand, and on the other some despondency. Why repeat the path already passed by someone earlier. It is better to fix your eyes on one of the already existing Unix-like systems, especially since they are all based on Unix Edition 7 and adapt it for your hardware.
I have a microcomputer P112 and there is an UZI180 for it. However, this operating system is not complete and there are errors in it. I really wanted a native compiler to appear in this system, on which its code is written. This prompted me to try to restore the source code from the HiTech C v3.09 compiler executables for CP/M and compile it for UZI180. In part, Mark Ogden and I have had some success. The source code of the compiler has been restored. A cross-compiler has been created from the recreated code, in which some errors present in the CP/M version have been fixed. The cross-compiler removes some of the restrictions on the size of the source modules and is fully compatible with the native compiler for CP/M. However, when compiling the restored code for UZI180, unexpected obstacles arose. Compiled executables do not work properly in UZI180. The same source code compiled for CP/M and for Linux, macOS or Windows works fine. This fact suggests that some of the functions of the UZI180 are implemented incorrectly and have errors. Unfortunately, many of them are included in the operating system kernel, which means that you need to change the entire operating system, including the kernel.
Changes in the kernel can lead to errors in other utilities that ensure the operation of the system as a whole. Adaptation and correction of errors in the UZI180 for the P112 microcomputer was handled by Hector Peraza. I am not sure that he will want to return to this project and start improving it and fixing bugs. At one time, he did a lot of work and somehow or poorly UZI180 works. There is not much in it yet.
I lack a simple text editor, a program like Kermit for sharing with other computers and a description of the utilities implemented in UZI180. I have already done a description of the kernel functions and almost all utilities. However, even here a fly in the ointment was found. Hector Peraza used previously written code by other programmers as the source code of utilities and compiled it into executable files. However, he could not physically check how they work. I found that some of them do not work correctly. My dream is not yet feasible. I don't believe yet that I can fix UZI180 alone and complete my plans.

Best regards,
Andrey
